pkg/client/oci/oci.go

```diff
@@ -62,18 +72,75 @@ func New(repo *api.Repo, c cache.Cacher, insecure bool) (*Repo, error) {
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
+	resolver := newDockerResolver(u, repo.GetAuth().GetUsername(), repo.GetAuth().GetPassword(), insecure)
+	// Init entries
+	entries := make(map[string][]string)
 
-	return NewRaw(u, repo.GetAuth().GetUsername(), repo.GetAuth().GetPassword(), c, insecure)
+	// First we try to load the index referenced from config file,
+	// if it does not exist, then we try to guess it directly from repo url
+	ociIndexRef := repo.GetOciIndex()
+	if ociIndexRef == "" {
+		klog.Infof("OCI index reference is empty in config file. Guessing index from repo url")
+		ociIndexRef = fmt.Sprintf("%s%s/index:latest", u.Host, u.Path)
+	}
+
+	ociIndexExists, err := ociReferenceExists(u, ociIndexRef, repo.GetAuth().GetUsername(), repo.GetAuth().GetPassword(), insecure)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	if ociIndexExists {
+		// Pull with ORAS lib
+		filename := "asset-index.json"
+		defer os.Remove(filename)
+		ctx := context.Background()
+		klog.Infof("Pulling index from %s...", ociIndexRef)
+		fileStore := content.NewFileStore("")
+		defer fileStore.Close()
+		_, _, err := oras.Pull(ctx, resolver, ociIndexRef, fileStore)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		// Read downloaded file
+		ociIndexBytes, err := ioutil.ReadFile(filename)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		r := bytes.NewReader(ociIndexBytes)
+
+		// Unmarshall index and populate entries
+		ociIndex := &api.OciIndex{}
+		err = pbjson.NewDecoder(r).Decode(ociIndex)
+		for _, c := range ociIndex.Charts{
+			for _, version := range c.Versions{
+				entries[c.Name] = append(entries[c.Name], version.Version)
+			}
+		}
+	} else {
+		klog.Info("Unable to load OCI index file. Using 'charts' filter from config file...")
+	}
+
+	return NewRaw(u, repo.GetAuth().GetUsername(), repo.GetAuth().GetPassword(), c, insecure, entries, resolver)
 }
 
 // NewRaw creates a Repo object.
-func NewRaw(u *url.URL, user string, pass string, c cache.Cacher, insecure bool) (*Repo, error) {
-	return &Repo{url: u, username: user, password: pass, cache: c, insecure: insecure}, nil
+func NewRaw(u *url.URL, user string, pass string, c cache.Cacher, insecure bool, entries map[string][]string, resolver remotes.Resolver) (*Repo, error) {
+	return &Repo{url: u, username: user, password: pass, cache: c, insecure: insecure, entries: entries, dockerResolver: resolver}, nil
 }
 
 // List lists all chart names in a repo
 func (r *Repo) List() ([]string, error) {
-	return nil, errors.Errorf("list method is not supported yet")
+	var names []string
+	// If entries is not populated, it means we couldn't load any index file so we need the charts filter in the
+	// configuration file
+	if len(r.entries) == 0 {
+		return nil, errors.New("No OCI index file detected. Please provide the charts filter in config file")
+	}
```

Comment:
I think this is out of the scope of this method's purpose. It is OK to return no items in a `#List` method. The caller must handle this case and perform an action accordingly. Actually, the OCI index is not a cause either since an index might be empty due to having no charts in the registry.

Reply:
Added some changes related to this. 

We were only calling to oci.List() method from index.go (loadCharts() method) if charts filter was not provided in the config file. So once we reach the oci.List() method, if entries is empty means we couldn't load any index file and so we have no idea what to sync. That's why I was returning the error at oci.List() method before

Now, oci.List() returns an empty list if entries is empty and the caller (loadCharts() methods) will check this specific case (source is OCI, no entries and no charts filter) and will throw the error. Note I added an abortExecution flag to finish charts-syncer execution depending on the error reported in loadCharts(). For some of them we want the execution to continue as they are just errors for an specific chart version.